File: ContentProviderRemoteDemo/app/src/main/java/edu/cs4730/contentproviderremotedemo/myAdapter.java
package edu.cs4730.contentproviderremotedemo;

import android.annotation.SuppressLint;
import android.database.Cursor;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.RecyclerView;

import edu.cs4730.contentproviderremotedemo.databinding.RecyclerRowBinding;

/**
 * this adapter is very similar to the adapters used for listview, except a ViewHolder is required
 * see http://developer.android.com/training/improving-layouts/smooth-scrolling.html
 * except instead having to implement a ViewHolder, it is implemented within
 * the adapter.
 */

public class myAdapter extends RecyclerView.Adapter<myAdapter.ViewHolder> {

    private Cursor cursor;

    private final String TAG = "myAdapter";
    private CursorViewModel mViewModel;

    //viewbinding provides the references now.
    public static class ViewHolder extends RecyclerView.ViewHolder {
        public RecyclerRowBinding viewBinding;

        public ViewHolder(RecyclerRowBinding viewBinding) {
            super(viewBinding.getRoot());
            this.viewBinding = viewBinding;
        }
    }

    //constructor
    public myAdapter(LifecycleOwner lifecycleOwner, CursorViewModel mv) {
        mViewModel = mv;
        //the observer could just as easily be in the MainActivity and the use a "swapCursor" method
        // to change the cursor and call notifyDataSetChanged()
        mViewModel.getData().observe(lifecycleOwner, new Observer<Cursor>() {
            @Override
            public void onChanged(Cursor data) {
                cursor = data;
                notifyDataSetChanged();
            }
        });
    }

    // Create new views (invoked by the layout manager)
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        RecyclerRowBinding v = RecyclerRowBinding.inflate(LayoutInflater.from(viewGroup.getContext()), viewGroup, false);
        return new ViewHolder(v);
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    @SuppressLint("Range")
    public void onBindViewHolder(ViewHolder viewHolder, int i) {
        if (setRow(i)) {  //move the cursor to the correct spot in the list other no data.
            String name = cursor.getString(cursor.getColumnIndex(MainActivity.KEY_NAME));
            String scoreStr = cursor.getString(cursor.getColumnIndex(MainActivity.KEY_SCORE));
            //note the key_score is int value, but we can request it as a string.

            viewHolder.viewBinding.name.setText(name);
            viewHolder.viewBinding.score.setText(scoreStr);

        }
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        if (cursor != null) {
            return cursor.getCount();
        } else {
            return 0;
        }
    }

    /**
     * @see android.widget.ListAdapter#getItemId(int)
     */
    @Override
    @SuppressLint("Range")
    public long getItemId(int position) {
        if (cursor != null) {
            if (cursor.moveToPosition(position)) {
                return cursor.getLong(cursor.getColumnIndex(MainActivity.KEY_ROWID));
            } else {
                return 0;
            }
        } else {
            return 0;
        }
    }

    /**
     * my class to find the correct row in the cursor
     */
    private boolean setRow(int position) {
        if (cursor != null) {
            return cursor.moveToPosition(position);
        }
        return false; //no data!
    }
}

File: ContentProviderRemoteDemo_kt/app/src/main/java/edu/cs4730/contentproviderremotedemo_kt/myAdapter.kt
package edu.cs4730.contentproviderremotedemo_kt

import android.annotation.SuppressLint
import android.database.Cursor
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import edu.cs4730.contentproviderremotedemo_kt.MainActivity.Companion.KEY_NAME
import edu.cs4730.contentproviderremotedemo_kt.MainActivity.Companion.KEY_ROWID
import edu.cs4730.contentproviderremotedemo_kt.MainActivity.Companion.KEY_SCORE
import edu.cs4730.contentproviderremotedemo_kt.databinding.RecyclerRowBinding

/**
 * this adapter is very similar to the adapters used for listview, except a ViewHolder is required
 * see http://developer.android.com/training/improving-layouts/smooth-scrolling.html
 * except instead having to implement a ViewHolder, it is implemented within
 * the adapter.
 */
@SuppressLint("NotifyDataSetChanged")
class myAdapter(lifecycleOwner: LifecycleOwner, private val mViewModel: CursorViewModel) :
    RecyclerView.Adapter<myAdapter.ViewHolder>() {
    private var cursor: Cursor? = null

    private val TAG = "myAdapter"

    //viewbinding provides the references now.
    class ViewHolder(var viewBinding: RecyclerRowBinding) : RecyclerView.ViewHolder(
        viewBinding.root
    )

    //constructor
    init {
        //the observer could just as easily be in the MainActivity and the use a "swapCursor" method
        // to change the cursor and call notifyDataSetChanged()
        mViewModel.data.observe(lifecycleOwner) { data ->
            cursor = data
            notifyDataSetChanged()
        }
    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        val v = RecyclerRowBinding.inflate(LayoutInflater.from(viewGroup.context), viewGroup, false)
        return ViewHolder(v)
    }

    // Replace the contents of a view (invoked by the layout manager)
    @SuppressLint("Range")
    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        if (setRow(i)) {  //move the cursor to the correct spot in the list other no data.
            val name = cursor!!.getString(cursor!!.getColumnIndex(KEY_NAME))
            val scoreStr = cursor!!.getString(cursor!!.getColumnIndex(KEY_SCORE))

            //note the key_score is int value, but we can request it as a string.
            viewHolder.viewBinding.name.text = name
            viewHolder.viewBinding.score.text = scoreStr
        }
    }

    // Return the size of your dataset (invoked by the layout manager)
    override fun getItemCount(): Int {
        return cursor?.count ?: 0
    }

    /**
     * @see android.widget.ListAdapter.getItemId
     */
    @SuppressLint("Range")
    override fun getItemId(position: Int): Long {
        return if (cursor != null) {
            if (cursor!!.moveToPosition(position)) {
                cursor!!.getLong(cursor!!.getColumnIndex(KEY_ROWID))
            } else {
                0
            }
        } else {
            0
        }
    }

    /**
     * my class to find the correct row in the cursor
     */
    private fun setRow(position: Int): Boolean {
        if (cursor != null) {
            return cursor!!.moveToPosition(position)
        }
        return false //no data!
    }
}
